### DISCUSSION TITLE
Set sample rate per error type

### DISCUSSION THREAD

**Discussion by zsofiaserenyi** [getsentry/sentry-cocoa#3211] [Q&A]
I'm currently using Sentry iOS SDK version `8.9.4` and I'm looking to transition from `Crashlytics` to `Sentry`. However, the quantity of custom errors (non-fatal errors) is quite high, making a direct 1:1 migration challenging. According to the information in the [documentation](https://docs.sentry.io/platforms/apple/guides/ios/configuration/sampling/#sampling-error-events), adjusting the `sampleRate` seems to impact all error types uniformly.

My **goal** is to **receive notifications for all crashes** without a percentage-based limitation, while still **applying a sampling rate** to the **custom errors** to manage their volume. Is there a method/best practice to accomplish this within the Sentry iOS SDK?🤔

**Comment by armcknight** [comment]
Hi @zsofiaserenyi, thanks for writing in, this is a good question. We have sampling mechanisms for three things, events, traces and profiles. For each, you can set a numerical sample rate. For traces/profiles, you can also set a custom block you can add arbitrary logic in where you could make such a decision (see `tracesSampler` and `profilesSampler`), but there's no equivalent API for event-level custom sampling logic like `eventsSampler` 🤔 This seems like it'd be a great feature addition, I'm going to ask about this with the team.

**Reply by armcknight** [reply]
@zsofiaserenyi Quick followup here, you should be able to get the same functionality by using SentryOptions.beforeSend, please check that out and let me know if it'd work for your purposes!

**Reply by zsofiaserenyi** [reply]
Hi Andrew, thank you very much for getting back to me so quickly, I appreciate your prompt help with my issue.:)

Based on the [documentation](https://docs.sentry.io/platforms/apple/configuration/filtering/?original_referrer=https%3A%2F%2Fdocsbot.ai%2F) and my current understanding `beforeSend` can be used to:
- add additional info to the event
- drop events 

So I could do something like this to **drop all non-fatal errors**:
```swift
  options.beforeSend = { event in
      event.level == .error ? nil : event
  }
```

However, this is **not the desired outcome** because I would like to **send some %** of the non-fatal errors.

So I could also do something like this to **only send 5%** of the non-fatal errors:
```swift
  options.beforeSend = { event in
      guard event.level == .error else { return event }
      return Int.random(in: 0 ... 100) <= 5 ? event : nil
  }
```

This means setting up a simple way to **select only some of the errors** **locally**.

I'm searching for a solution that **doesn't require** such logic from my side as my assumption is that the `sampleRate` option within the SDK possesses a **more sophisticated capability** than what I can achieve through the above method. 🤔

Let me know **if I've overlooked** any facets of the `beforeSend` functionality and if there are **other ways** to achieve my goal.🙏

**Reply by philipphofmann** [reply]
What you described with 
```Swift
  options.beforeSend = { event in
      guard event.level == .error else { return event }
      return Int.random(in: 0 ... 100) <= 5 ? event : nil
  }
```
is the current way to go. 

As pointed out by @armcknight there is no API for custom sampling of events. @armcknight, did you get any feedback on such an API? I agree with you that such an API could make sense.

**Reply by acfriedman** [reply]
Hi @philipphofmann 

I'd suspect we would need to set the `sampleRate` to `1.0` if we do this custom sampling with `beforeSend`? If not, we would be applying a sample to a sample, right?

**Reply by philipphofmann** [reply]
Sorry for the late reply, @acfriedman; I was on PTO for a month. Yes, indeed, you need to set the sampleRate to 1.0 for your use case.
